Parse `noSolution` status from CPLEXShell and CPLEXDirect solvers correctly

ruaridhw opened the pull request:
## Fixes

## Summary/Motivation:
Currently when CPLEX returns "No Solution", this isn't captured in any of the Pyomo solution or solver data objects. We can identify "No Solution" because when both of (1) max time limit is reached and (2) the model is infeasible occur.

## Changes proposed in this PR:

### `CPLEXDirect`
- Clean up the status codes in `_postsolve()` to use the actual CPLEX library constants
- Add new logic to handle `MIP_time_limit_infeasible` and `MIP_dettime_limit_infeasible` as indicating a solver `termination_condition` of `noSolution`
- Get the solver return code if possible from any `CplexSolverError`s (but don't add this to the returned `Bunch` due to downstream code raising `ApplicationError`s)
- Set `results.solver.return_code` and `results.solver.termination_message` using the available information

### `CPLEXSHELL`
- Fix `process_logfile()` for CPLEX 12.10 as it looks like IBM have updated the log file schema and is no longer being parsed correctly
- Add new logic to handle `'no integer solution.'` in the log file as indicating a solver `termination_condition` of `noSolution`
- Get the solver return code if possible from any `CPLEX Error` lines in the log file

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Commit:
:bug: Fix `process_logfile()` for CPLEX 12.10

- It looks as though IBM have updated the schema that the shell logfile uses and therefore the parsing is not quite done correctly
- Notably, 'CPLEX Error; no integer solution' is being flagged and returned as ('ok', 'unknown') for the status and termination condition

Commit:
:books: Use the `status` enum instead of magic numbers

Commit:
:hammer: Return `noSolution` for Direct solver to match Shell

- The Direct solver should return the same solver and solution status as the Shell solver when no solution is found
- These error codes are currently unhandled and just return `TerminationCondition.error`
- It seems reasonable to assume that if a solve reaches max time limit and is infeasible that we cannot yet determine whether it is actually infeasible and therefore makes sense to return `noSolution`.

Commit:
:hammer: Update `termination_message`

Commit:
:hammer: Extract the solver `return_code`s should they exist

- I've elected not to add this to the return Bunch from `_apply_solver()` because that would change the behaviour of the downstream code (which raises an ApplicationError)

Commit:
:rotating_light: Add tests of `noSolution` status

ikonikon commented on pyomo/solvers/plugins/solvers/cplex_direct.py:
```diff
@@ -180,9 +180,15 @@ def _process_stream(arg):
                     opt_cmd.set(float(option))
-            
+
+            self._rc = None
```
This is only populated in cases of a raised `CplexSolverError` and will be `None` in cases of successful solves. Are we ok with that? I would have thought we should initialise it with the return code for a successful run instead (if it exists)

ruaridhw replied:
Yeah good point. I presumed that the return code would always be `0` for a successful run and that any downstream code wouldn't find that particularly interesting. Will update anyway

ruaridhw replied:
Actually, that's not quite correct. It **should** only be populated in case of a `CplexSolverError` because its actually misnamed; `self._error_code` is its true purpose. `Cplex.solve()` always returns `None` so there's no return code as such. Only an error code and a solution status code.

ikonikon replied:
OK, shall we rename to `self._error_code` then? Up to you really, i do not mind.

Commit:
:books: Rename 'return code' to 'error code'

- Do not adjust `SolverInformation` since this is called `return_code` here

Commit:
:hammer: Expand the criterion for `noSolution`

- Reaching these limits could also be reasons for having not found a solution

Commit:
:rotating_light: Fix test to make the model more unlikely to not solve

Commit:
:rotating_light: Add test of presolve infeasible

Commit:
:bug: Set `SolverStatus.warning` instead of `error` for reasonable termination

- If we have already determined a reasonable termination condition that would have otherwise returned an `ok` SolverStatus, return `warning` instead of `error`

Commit:
:rotating_light: Fix nosolution test

Commit:
:bug: Update Direct solver to behave the same as Shell solver

- In the case of `noSolution`, CPLEXSHELL now considers this a `warning` solver status
- Similarly, `GAMSDirect` returns a `SolutionStatus.unknown` so we should be consistent with that

Commit:
:rotating_light: Update test

- Constraint `C` is being overwritten

Commit:
:hammer: Refactor test to use MTZ TSP

- This is a much harder model that is more likely to not be solved within the allowed time
